anvil-python/anvil/commands/upgrades/intra_channel.py

```diff
@@ -158,40 +158,44 @@ def get_plan(self) -> list[BaseStep]:
         plan: list[BaseStep] = []
         plan.append(LatestInChannel(self.jhelper, self.manifest))
 
-        plan.extend(
-            haproxy_upgrade_steps(
-                self.client,
-                self.manifest,
-                self.jhelper,
-                self.deployment.infrastructure_model,
-                self.preseed,
+        if self.client.cluster.list_nodes_by_role("haproxy"):
+            plan.extend(
+                haproxy_upgrade_steps(
+                    self.client,
+                    self.manifest,
+                    self.jhelper,
+                    self.deployment.infrastructure_model,
+                    self.preseed,
+                )
             )
-        )
-        plan.extend(
-            postgresql_upgrade_steps(
-                self.client,
-                self.manifest,
-                self.jhelper,
-                self.deployment.infrastructure_model,
-                self.preseed,
+        if self.client.cluster.list_nodes_by_role("database"):
```

Comment:
cluster has at least one database, otherwise we cannot have maas. As such this check is redundant and can be omitted.